DOC: Add warning for newbs not to edit auto-generated file

Garrett-R opened the pull request:
@jreback, here is the proposed changed we talked about in [this PR](https://github.com/pydata/pandas/pull/10337#issuecomment-111822584) for other fellow newbs that make the same mistake I did.

You said "and more importantly prob a note in `internals.rst` about how/what to change", but I didn't understand that.  `internals.rst` is not auto-generated, right?  So, why would I add a note there?  Perhaps you wanted a note to appear in `internals.html`?  That seems unnecessary since if the user has that file on their machine, they must have intentionally generated it...

Garrett-R committed:
DOC: Add warning for newbs not to edit auto-generated file

shoyer commented:
This is where `internals.rst` ends up in the generated docs: http://pandas.pydata.org/pandas-docs/stable/internals.html

That said, I think the warning in the generated file serves this purpose more directly. Warnings in docs tend to get ignored anyways :).

Garrett-R commented:
@shoyer, hmmm.. maybe I'm confused.  So are you saying it would be good to have a comment in `internals.html` warning new devs that `internals.html` is auto-generated and they should edit `internals.rst` instead?

If that's the case, we should probably have a similar comment in all the [generated] HTML files, right?

shoyer commented:
@Garrett-R No, I think @jreback was thinking you could have the warning "generated.pyx is generated from make_generated.py" in the internals documentation. None of the documentation HTML pages are checked into git, so there's no danger we could get a pull request editing them directly.

HOWEVER, I actually disagree with @jreback on this one. I think the warning you have here is sufficient.

Garrett-R commented:
Ah, I see.  Yeah, I agree with you.  I also think having the source code warning in `generated.pyx` should be enough.  It's already kind of embarrassing that I didn't realize that `generated.pyx` was generated.

jreback commented:
no the issue is that I don't think it's obvious that you need to Python generate_code.py

that instruction should live maybe in the contributing docs as well as in generated.pyx

Garrett-R commented:
Oh yeah, that's true, it's definitely not obvious that you need to run `python2 generatre_code.py` (I had to be told).  

The only situation you'd ever need to know this is if you're actually trying to modify `generated.pyx`, so one option would be to just include it these instructions in both `generated.pyx` and `generate_code.py` (which I've now done), but not add extra noise to the contributing docs.

jreback commented on pandas/src/generate_code.py:
```diff
@@ -1,3 +1,10 @@
+during building.  To regenerate `generated.pyx`, just run:
+
+    `python2 generate_code.py`.
+
```
just say python

Garrett-R replied:
You sure?  Some systems now have `python` pointing to python3 (ex: Arch Linux) whereas I believe all systems recognize `python2`... and this file is not Python 3 compatible...

shoyer replied:
+1 on python2.

On Sat, Jun 27, 2015 at 5:48 PM, Garrett Reynolds <notifications@github.com>
wrote:

> In pandas/src/generate_code.py
> <https://github.com/pydata/pandas/pull/10456#discussion_r33417787>:
>
> > @@ -1,3 +1,10 @@
> > +"""This file generates `generated.pyx` which is then included in `../algos.pyx`
> > +during building.  To regenerate `generated.pyx`, just run:
> > +
> > +    `python2 generate_code.py`.
> > +
>
> You sure? Some systems now have python pointing to python3 (ex: Arch
> Linux) whereas I believe all systems recognize python2... and this file
> is not Python 3 compatible...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/10456/files#r33417787>.
>

jreback commented:
hmm this should be compat but I guess that's another issue

ok then on the py2

Garrett-R commented:
Yeah, it would be nice to have it compatible.

jreback commented on pandas/src/generate_code.py:
```diff
@@ -1,3 +1,10 @@
+"""This file generates `generated.pyx` which is then included in `../algos.pyx`
+during building.  To regenerate `generated.pyx`, just run:
```
this actually has to be run from ``pandas/src`` (so that ``generated.pyx`` is in the correct place), so either we need to fix this (e.g. the output path should be relative to the pandas tree), or provide this in the instructions.

Garrett-R replied:
Right, I fixed it so it can now be run from anywhere.

Garrett-R committed:
Make generate_code.py Py3-friendly

Garrett-R committed:
Allow running `generate_code.py` from anywhere

Previously you had to run it from inside the same directory, but
it's less likely to mess up newbs if you can run it from anywhere.

Garrett-R commented on pandas/src/generate_code.py:
```diff
@@ -1,12 +1,23 @@
-# or we get a bootstrapping problem
-from StringIO import StringIO
+import os
+from pandas.compat import StringIO
```
I didn't understand this warning here and [the original commit](https://github.com/pydata/pandas/commit/21dabea425e1ee872fcfcaa36ecde991eb8d3a27) had no more information about this.  I tested this on Py2 and Py3 and it seemed to work fine...

jreback commented:
merged via https://github.com/pydata/pandas/commit/758ca05e2eb04532b5d78331ba87c291038e2c61

thanks!
